### PR TITLE
change: Update package metadata

### PR DESCRIPTION
**Pull request by ofek**
## Background

Hello there! The Python packaging ecosystem has standardized on the interface for build backends ([PEP 517](https://peps.python.org/pep-0517/)/[PEP 660](https://peps.python.org/pep-0660/)) and the format for metadata declaration ([PEP 621](https://peps.python.org/pep-0621/)/[PEP 631](https://peps.python.org/pep-0631/)). As a result, the execution of `setup.py` files is now [deprecated](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html).

So, I'm spending my free time updating important projects so that they are modernized and set an example for others 😄

## Summary of changes

This implements [PEP 621](https://peps.python.org/pep-0621/), obviating the need for `setup.py`, `setup.cfg`, and `MANIFEST.in`. The build backend `hatchling` (of which I am a maintainer in the [PyPA](https://github.com/pypa)) is now used as that is the default in the official [Python packaging tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/). Hatchling is available on all the major distribution channels such as [Debian](https://salsa.debian.org/python-team/packages/hatchling), [Fedora](https://src.fedoraproject.org/rpms/python-hatchling), [Arch Linux](https://archlinux.org/packages/community/any/python-hatchling/), [conda-forge](https://anaconda.org/conda-forge/hatchling), [Nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/hatchling/default.nix), [Alpine Linux](https://github.com/alpinelinux/aports/blob/master/community/py3-hatchling/APKBUILD), [FreeBSD](https://github.com/freebsd/freebsd-ports/blob/main/devel/py-hatchling/Makefile)/[OpenBSD](https://github.com/openbsd/ports/blob/master/devel/py-hatchling/Makefile), [Gentoo Linux](https://github.com/gentoo/gentoo/tree/master/dev-python/hatchling), [MacPorts](https://github.com/macports/macports-ports/blob/master/python/py-hatchling/Portfile), [OpenEmbedded](https://github.com/openembedded/openembedded-core/commit/846e806181f1349be29cbce78c5041735dfd7e6f), [Spack](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-hatchling/package.py), [MSYS2](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-python-hatchling/PKGBUILD), etc.

## Notes

- The source distributions on PyPI are shipping `/requirements/tox` even though `/tox.ini` is not; let me know if you want me to remove the former or add the latter.
- The project is no longer polluted with build artifacts such as a `*.egg-info` directory.
- (See ["Why Hatch?"](https://hatch.pypa.io/latest/why/#build-backend)) The project now supports static analysis for external tools such as IDEs. With setuptools, you must provide [additional configuration](https://setuptools.pypa.io/en/latest/userguide/development_mode.html#legacy-behavior) for editable installations which means that by default, for example, you would not get autocompletion in Visual Studio Code. This is marked as a legacy feature and may in fact be removed in future versions of setuptools.

## Testing done

```
tox -e py39
```

#### General

- [X] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [X] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [X] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [X] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by navinsoni — COMMENTED** [review]
/bot run all

**Comment by navinsoni** [comment]
@ofek Thank you for contributing to sagemaker-python-sdk. I will review it and get back to you.

**Comment by ofek** [comment]
Thanks! Would you mind triggering the CI again?

**Comment by ofek** [comment]
@navinsoni I rebased; would you mind triggering the CI again?

**Review by mufaddal-rohawala — COMMENTED** [review]
/bot run all

**Comment by ofek** [comment]
Would you mind triggering the CI again?

**Review by mufaddal-rohawala — COMMENTED** [review]
/bot run all

**Comment by mufaddal-rohawala** [comment]
This PR would need a change in build infra of the repo to make the move to new setup.

**Comment by ofek** [comment]
Oh okay, can you direct me to what additional files I should modify?

**Comment by mufaddal-rohawala** [comment]
Ok so this is internal codebase managed by Sagemaker which hosts logic for Codebuild Projects used in PR checks. They need to be modified and I have created an item in teams backlog for the same, once the item is complete we can resume/merge this PR.

**Comment by ofek** [comment]
Okay thanks!

**Comment by jmahlik** [comment]
It would be really awesome if a wheel got produced as a side effect of `hatch build` from https://github.com/aws/sagemaker-python-sdk/issues/3181

**Comment by ofek** [comment]
I rebased, if somebody could please allow the tests to run.

**Comment by ofek** [comment]
cc @liujiaorr @evakravi @gwang111 perhaps, I'm not sure who owns the repository maintenance anymore

I left some time for [this](https://github.com/aws/sagemaker-python-sdk/pull/3529#issuecomment-1504129156) to take place. Specifically, I set my reminder for 13 months because I know how long things can take!

I am committed to continue helping out with this endeavor 🙂

**Comment by knikure** [comment]
@ofek Thanks for our contribution. I have triggered CI on your PR.

**Comment by ofek** [comment]
Thank you! It looks like the AWS actions are trying to invoke `setup.py` directly which is deprecated as I mentioned in the PR description.

What should I do?

**Comment by ofek** [comment]
Could someone please provide an update on this?

**Comment by benieric** [comment]
> Could someone please provide an update on this?

Oh, thanks let me check on this updating to modernized packaging would be ideal. We need to make an infra change so our uses `python -m build` instead of setup.py method

**Comment by benieric** [comment]
There are also a few other places where we need to update to `python -m build` in this PR

https://github.com/aws/sagemaker-python-sdk/blob/b96c98e727e4150912445889a191b0e88517dad8/tests/integ/sagemaker/experiments/conftest.py#L158

https://github.com/search?q=repo%3Aaws%2Fsagemaker-python-sdk%20python3%20setup.py&type=code

https://github.com/aws/sagemaker-python-sdk/blob/b96c98e727e4150912445889a191b0e88517dad8/tests/integ/sagemaker/feature_store/feature_processor/test_feature_processor.py#L1275

**Comment by ofek** [comment]
I already accounted for the first one but I just modified the other two, thanks Erick! Do you know how challenging the infra modification you mentioned would be?

**Comment by benieric** [comment]
> I already accounted for the first one but I just modified the other two, thanks Erick! Do you know how challenging the infra modification you mentioned would be?

Should be pretty simple, but need to coordinate the merge a bit

**Comment by benieric** [comment]
Hey @ofek can we actually do something like this https://packaging.python.org/en/latest/guides/modernize-setup-py-project/#what-if-something-that-can-not-be-changed-expects-a-setup-py-file

move build-system to pyproject.toml but keep the setup.py file for process that expect it

**Comment by ofek** [comment]
I can do that but then you lose all of the benefits of moving to Hatchling. I could move to Hatchling in the new static metadata and then keep the `setup.py` file unchanged but then you would have to maintain both.

**Comment by benieric** [comment]
> I can do that but then you lose all of the benefits of moving to Hatchling. I could move to Hatchling in the new static metadata and then keep the `setup.py` file unchanged but then you would have to maintain both.
> 
> edit: also as I mentioned in the original description, execution of that file will simply not work eventually

Ah, I see, would be better to remove the setup.py, but I think I need to bring up with team before we merge this. As seen by some of our tests there may be workflows where customers are using `python setup.py` to create distribution of pysdk in order to upload in their remote jobs

**Comment by benieric** [comment]
Or we need to use setuptools as build system for pyproject.toml 
and keep a minimal setup.py like in the docs 

```
import setuptools

setuptools.setup()

```

**Comment by ofek** [comment]
I think I see a way to have the best of both worlds, I'll update tonight.

**Comment by ofek** [comment]
I added back the `setup.py` file and a deprecation warning will now display upon execution. The package data is now sourced from the standard mechanisms and I removed unnecessary metadata because that file will not be used to build the release artifacts.

**Comment by benieric** [comment]
I think might need to add `hatch_build.py` to `MANIFEST.in`?

**Comment by ofek** [comment]
That's not necessary because the legacy installation mechanism doesn't build a source distribution. I can add that for your peace of mind though, one sec.

**Comment by benieric** [comment]
Also, thanks alot, was also looking at your notes that say `The source distributions on PyPI are shipping /requirements/tox even though /tox.ini is not; let me know if you want me to remove the former or add the latter.` Can we remove the `/requirements/tox/` if it is simple?

**Comment by ofek** [comment]
Yes that's removed already, the source distribution contents are now defined by the following block:

```toml
[tool.hatch.build.targets.sdist]
only-include = [
  "/requirements/extras",
  "/src",
  "/VERSION",
]
```

in addition to what is defined in this note callout https://hatch.pypa.io/latest/plugins/builder/sdist/#default-file-selection

**Comment by ofek** [comment]
This appears to be passing now however there are stylistic errors corresponding to files that were not changed. I was going to merge the latest master but you've already done that so I'm not sure what to do.

**Comment by ofek** [comment]
The good news is that the proposed style change is not erroneous! The bad news is that somehow it's not doing the right thing on master (and other PRs).

**Comment by benieric** [comment]
Yeah lol, not sure why the style change issue is not being shown in master branch

**Comment by benieric** [comment]
Looks like the proposed style changes are correct in this PR, can we do those and include with this change? For black should be able to just do `tox -e black-format`

**Comment by ofek** [comment]
Done!

**Comment by ofek** [comment]
Since Python 2 is no longer supported by this project you can remove the requirement to add a `from __future__ import absolute_import` at the top of every file. In a case, I fixed that error and I think if you rerun CI everything should pass now!

File: tox.ini
Comment (benieric) [inline tox.ini:52]: doc8, is complaining about some file in `src/sagemaker.egg-info` may need to add this part back

Comment (ofek) [reply tox.ini]: Done, thanks!

**Review by benieric — APPROVED** [review]

**Comment by ofek** [comment]
Looks like a flaky test, might just need to rerun that particular failed job.

**Comment by benieric** [comment]
Thanks for working on this @ofek , will get one more review from team member before we get this merged!

**Comment by ofek** [comment]
Hello again, I just wanted to follow-up to see how things are going!

**Comment by benieric** [comment]
Sorry for the delay, I am reaching out to team member today to get a second review!

**Comment by benieric** [comment]
@ofek Could you take a look at the conflicts please. @akrishna1995 has agreed to help review this one

**Comment by ofek** [comment]
Sure, just updated the branch!

**Comment by benieric** [comment]
Thanks, sorry again for the churn

File: pyproject.toml
Comment (benieric) [inline pyproject.toml:10]: Just saw this but can you change this to be `license = {file = "LICENSE.txt"}` like mentioned [here](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license)

Comment (ofek) [reply pyproject.toml:10]: That is outdated actually the standard is being updated by [PEP 639](https://peps.python.org/pep-0639/) which was always planned (allowing a string value). The license file is automatically picked up by Hatchling and included in the source distribution.

**Review by mufaddal-rohawala — DISMISSED** [review]

File: pyproject.toml
Comment (benieric) [reply pyproject.toml:10]: The only blocker I have with this one currently is that when I run `python setup.py sdist` locally I get an error like:

```
configuration error: `project.license` must be valid exactly by one definition (2 matches found):

    - keys:
        'file': {type: string}
      required: ['file']
    - keys:
        'text': {type: string}
      required: ['text']

DESCRIPTION:
    `Project license <https://peps.python.org/pep-0621/#license>`_.

GIVEN VALUE:
    "Apache-2.0"

OFFENDING RULE: 'oneOf'
```

Comment (benieric) [reply pyproject.toml:10]: Because [PEP 639](https://peps.python.org/pep-0639/) is still in Draft can we do `license = {file = "LICENSE.txt"}` until then?

Comment (ofek) [reply pyproject.toml:10]: Done! I removed the field instead because the license is automatically added to the source distribution.

**Review by benieric — APPROVED** [review]

**Comment by ofek** [comment]
Thanks for the perseverance here! I will start helping your organization migrate other projects in my free time 🙂

**Pull request merged**
